### PR TITLE
Sol/lp strategy

### PR DESCRIPTION
**Pull request by Sol3iman**
- Added LPOptimiser Strategy contract
- Rebalance check functions verifies whether the current fixed rate is outside of the current position's tick range
- Rebalance function calls the rebalanceCheck function and calculates new lower and upper ticks, converts them to multiples of desired tickRange and emits an event when the rebalance is done with the new tickLower and tickUpper.
- For now information about the ticks and current rates are mocked in the tests

### CODE REVIEW

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 1]
```diff
@@ -6,11 +6,14 @@ import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
...
 import "../interfaces/external/voltz/IMarginEngine.sol";
+import "../interfaces/external/voltz/IPeriphery.sol";
+import "../interfaces/external/voltz/IVAMM.sol";
 import "../interfaces/utils/ILpCallback.sol";
 import "../libraries/ExceptionsLibrary.sol";
 import "../libraries/CommonLibrary.sol";
 import "../libraries/external/FullMath.sol";
 import "../utils/DefaultAccessControl.sol";
+import "@prb/math/contracts/PRBMathSD59x18.sol";
 import "hardhat/console.sol";
```
Comment (arturbeg) [inline contracts/strategies/LPOptimiserStrategy.sol:16]: gotta remove this from PRs

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 2]
```diff
@@ -25,12 +28,23 @@ contract LPOptimiserStrategy is DefaultAccessControl, ILpCallback {
     IVoltzVault internal _vault;
     uint256[] internal _pullExistentials;
     IMarginEngine internal _marginEngine;
+    IPeriphery internal _periphery;
+    IVAMM internal _vamm;
     IVoltzVault.TickRange internal _currentPosition;
 
     // MUTABLE PARAMS
 
     uint256 _lastSignal;
```
Comment (arturbeg) [inline contracts/strategies/LPOptimiserStrategy.sol]: are the below variables public?

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 3]
```diff
@@ -25,12 +28,23 @@ contract LPOptimiserStrategy is DefaultAccessControl, ILpCallback {
...
     uint256 _lastLeverage;
+    uint256 _sigmaWad; // y (standard deviation parameter in wad 10^18)
+    uint256 _max_possible_lower_bound; // should be in fixed rate
+    uint256 _k_unwind_parameter; // parameter for k*leverage (for unwinding so this needs to be sent to the contract vault but not used in the strategy vault)
```
Comment (arturbeg) [inline contracts/strategies/LPOptimiserStrategy.sol]: need more informative comments in here @0xZenus @Sol3iman, g is the value in wad then should always have wad in the variable name for simplicity

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 4]
```diff
@@ -43,11 +57,133 @@ contract LPOptimiserStrategy is DefaultAccessControl, ILpCallback {
...
         _vault = vault_;
         _marginEngine = IMarginEngine(vault_.marginEngine());
+        _periphery = IPeriphery(vault_.periphery());
+        _vamm = IVAMM(vault_.vamm());
         _currentPosition = vault_.currentPosition();
         _tokens = vault_.vaultTokens();
         _pullExistentials = vault_.pullExistentials();
     }
 
+    event rebalanced(int24 newTickLowerMul, int24 newTickUpperMul);
```
Comment (arturbeg) [inline contracts/strategies/LPOptimiserStrategy.sol]: do you mean emit instead of event?

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 5]
```diff
@@ -43,11 +57,133 @@ contract LPOptimiserStrategy is DefaultAccessControl, ILpCallback {
...
+    function setConstants(int24 logProx, uint256 sigmaWad, uint256 max_possible_lower_bound, int24 tickSpacing) public {
+        _requireAtLeastOperator();
+        _logProximity = logProx;
+        _sigmaWad = sigmaWad;
+        _max_possible_lower_bound = max_possible_lower_bound;
+        _tickSpacing = tickSpacing;
+    }
+
+    function setTickValues(int24 currentTick, int24 tickLower, int24 tickUpper) public {
+        _requireAtLeastOperator();
```
Comment (arturbeg) [inline contracts/strategies/LPOptimiserStrategy.sol:120]: is there no modifier for this @0xZenus @Sol3iman ?

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 6]
```diff
@@ -43,11 +57,133 @@ contract LPOptimiserStrategy is DefaultAccessControl, ILpCallback {
...
+    /// @notice Get the current tick and position ticks and decide whether to rebalance
+    function rebalanceCheck() public view returns (bool) {
+        
+        // 1. Get current position, lower, and upper ticks
+        // _currentPosition = _vault.currentPosition();
+        // int24 _tickLower = _currentPosition.tickLower;
+        // int24 _tickUpper = _currentPosition.tickUpper;
+
+        // 2. Get current tick
+        // int24 _currentTick = _periphery.getCurrentTick(_periphery.address());
```
Comment (arturbeg) [inline contracts/strategies/LPOptimiserStrategy.sol]: why would we pass _periphery.address() into .getCurrentTick(...) ?

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 7]
```diff
@@ -43,11 +57,133 @@ contract LPOptimiserStrategy is DefaultAccessControl, ILpCallback {
...
+
+        // 3. Compare current fixed rate to lower and upper bounds
+        if (
+            _tickLower - _logProximity <= _currentTick &&
```
Comment (arturbeg) [inline contracts/strategies/LPOptimiserStrategy.sol]: @0xZenus will be helpful to have a short writeup of how the math works when using logs, i assume the same and it's more gas efficient

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 8]
```diff
@@ -43,11 +57,133 @@ contract LPOptimiserStrategy is DefaultAccessControl, ILpCallback {
...
+            // 4.1. If current fixed rate is within bounds, return false (don't rebalance)
+            return (false);
+        } else {
+            // 4.2. If current fixed rate is outside bounds, return true (do rebalance)
+            return (true);
+        }
+    }
+
+    /// @notice Get the nearest tick multiple given a tick and tick spacing
+    function nearestTickMultiple(int24 newTick, int24 tickSpacing) internal pure returns (int24) {
```
Comment (arturbeg) [inline contracts/strategies/LPOptimiserStrategy.sol]: is there a test for this function?

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 9]
```diff
@@ -43,11 +57,133 @@ contract LPOptimiserStrategy is DefaultAccessControl, ILpCallback {
...
+     return (newTick / tickSpacing + (newTick % tickSpacing >= tickSpacing/2 ? int24(1) : int24(0)) ) * tickSpacing;
+    }
+
+    /// @notice Set new optimimal tick range based on current tick
+    /// @param currentFixedRateWad currentFixedRate which is passed in from a 7-day rolling avg. historical fixed rate.
+    function rebalance (int256 currentFixedRateWad) public returns (int256 newTickLower, int256 newTickUpper) {
+        _requireAtLeastOperator();
+
+        // console.log('my console logs start here:');
```
Comment (arturbeg) [inline contracts/strategies/LPOptimiserStrategy.sol]: will be good to clean this up for the final merge or convert these granular steps into unit tests?

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 10]
```diff
@@ -43,11 +57,133 @@ contract LPOptimiserStrategy is DefaultAccessControl, ILpCallback {
...
+        // console.log(PRBMathUD60x18.mul(_sigmaWad, 2000000000000000000));
+        // console.logInt(nearestTickMultiple(_newTickLower, _tickSpacing));
+
+        if (rebalanceCheck()) {
+            // 0. Get tickspacing from vamm
+            // int24 _tickSpacing = _vamm.tickSpacing(_vamm.address());
+
+            // 1. Get the new tick lower
+            // uint256 _newFixedLowerWad = Math.min(Math.max(0, uint256(currentFixedRateWad) - _sigmaWad), _max_possible_lower_bound);
+            uint256 deltaWad = uint256(currentFixedRateWad) - _sigmaWad;
```
Comment (arturbeg) [inline contracts/strategies/LPOptimiserStrategy.sol]: deltaWad has to be an int256 rather than a uint256, wdyt @0xZenus ?

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 11]
```diff
@@ -43,11 +57,133 @@ contract LPOptimiserStrategy is DefaultAccessControl, ILpCallback {
...
+            if (deltaWad > 0) {
+                // delta is greater than 0 => choose delta
+                if (deltaWad < _max_possible_lower_bound) {
+                    _newFixedLowerWad = deltaWad;
+                } else {
+                    _newFixedLowerWad = _max_possible_lower_bound;
+                }
+            } else {
+                // delta is less than 0 => choose 0
+                if (_max_possible_lower_bound > 0) {
```
Comment (arturbeg) [inline contracts/strategies/LPOptimiserStrategy.sol]: will be good to have comments to explain this logic

Comment (arturbeg) [reply contracts/strategies/LPOptimiserStrategy.sol]: i don't think _max_possible_lower_bound can ever be negative, so the logic feels redundant?

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 12]
```diff
@@ -43,11 +57,133 @@ contract LPOptimiserStrategy is DefaultAccessControl, ILpCallback {
...
+                } else {
+                    _newFixedLowerWad = _max_possible_lower_bound;
+                }
+            }
+            // 2. Get the new tick upper
+            console.log(_newFixedLowerWad);
+            uint256 _newFixedUpperWad = _newFixedLowerWad + 2 * _sigmaWad;
+            console.log(_newFixedUpperWad);
+            // 3. Convert new fixed lower rate back to tick
+            int256 _newTickLowerWad = -PRBMathSD59x18.div(PRBMathSD59x18.log2(int256(_newFixedUpperWad)),
```
Comment (arturbeg) [inline contracts/strategies/LPOptimiserStrategy.sol]: need to be careful so that this value is not below MIN TICK, with a smart selection of _max_possible_lower_bound in theory we can avoid this scenario but still feels worth having an extra check @0xZenus @Sol3iman ?

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 13]
```diff
@@ -43,11 +57,133 @@ contract LPOptimiserStrategy is DefaultAccessControl, ILpCallback {
...
+            console.log(_newFixedLowerWad);
+            uint256 _newFixedUpperWad = _newFixedLowerWad + 2 * _sigmaWad;
+            console.log(_newFixedUpperWad);
+            // 3. Convert new fixed lower rate back to tick
+            int256 _newTickLowerWad = -PRBMathSD59x18.div(PRBMathSD59x18.log2(int256(_newFixedUpperWad)), 
+                                                        PRBMathSD59x18.log2(1000100000000000000)
+                                                        );
+            console.logInt(_newTickLowerWad);
+            // 4. Convert new fixed upper rate back to tick
+            int256 _newTickUpperWad = -PRBMathSD59x18.div(PRBMathSD59x18.log2(int256(_newFixedLowerWad)),
```
Comment (arturbeg) [inline contracts/strategies/LPOptimiserStrategy.sol]: what's the reasoning behind using prb math for this conversion instead of doing it similar to this https://github.com/Voltz-Protocol/voltz-core/blob/b3b66a25bdbef0b36e786cabe5a912e3fff2a3b8/contracts/MarginEngine.sol#L156 but in the opposite direction @0xZenus @Sol3iman ?

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 14]
```diff
@@ -43,11 +57,133 @@ contract LPOptimiserStrategy is DefaultAccessControl, ILpCallback {
...
+            console.logInt(_newTickLower);
+            console.logInt(_newTickUpper);
+
+            int24 _newTickLowerMul = nearestTickMultiple(int24(_newTickLower), _tickSpacing);
+            int24 _newTickUpperMul = nearestTickMultiple(int24(_newTickUpper), _tickSpacing);
+
+            console.logInt(_newTickLowerMul);
+            console.logInt(_newTickUpperMul);
+
+            emit rebalanced(_newTickLowerMul, _newTickUpperMul);
```
Comment (arturbeg) [inline contracts/strategies/LPOptimiserStrategy.sol]: would name the event Rebalanced with capital to be aligned with the standards

File: contracts/strategies/LPStrategyIndex.ts [hunk 15]
```diff
@@ -0,0 +1,69 @@
+import { ethers } from "ethers";
```
Comment (arturbeg) [inline contracts/strategies/LPStrategyIndex.ts:1]: is this the script that will be used by OZ to trigger rebalances?

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 5 above)
Comment (0xZenus) [reply contracts/strategies/LPOptimiserStrategy.sol:120]: _requireAtLeastOperator does the job. (here's the implementation : `require(
            governance.isOperator(msg.sender) || governance.isAdmin(msg.sender), 
            ExceptionsLibrary.FORBIDDEN
        );`.
        
However, there's also _requireAdmin. We need to ask Mellow who is operator and who is admin and then decide how to change.

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 4 above)
Comment (0xZenus) [reply contracts/strategies/LPOptimiserStrategy.sol]: I think here's where the event is declared.

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 6 above)
Comment (0xZenus) [reply contracts/strategies/LPOptimiserStrategy.sol]: We should pass _marginEngine.address() into .getCurrentTick(...)

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 7 above)
Comment (0xZenus) [reply contracts/strategies/LPOptimiserStrategy.sol]: Sure. https://www.notion.so/voltz/Proximity-parameter-ed5326ebd7d24d2f885e3912b105ac59

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 8 above)
Comment (0xZenus) [reply contracts/strategies/LPOptimiserStrategy.sol]: The logic seems good. However, `newTick % tickSpacing >= tickSpacing/2` might encounter issues when newTick is negative. It's worth adding UTs to this function and cover cases where newTick is both negative and positive.

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 9 above)
Comment (0xZenus) [reply contracts/strategies/LPOptimiserStrategy.sol]: Agree.

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 10 above)
Comment (0xZenus) [reply contracts/strategies/LPOptimiserStrategy.sol]: Definitely.

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 16]
```diff
@@ -43,11 +57,133 @@ contract LPOptimiserStrategy is DefaultAccessControl, ILpCallback {
...
+    }
+
+    /// @notice Get the nearest tick multiple given a tick and tick spacing
+    function nearestTickMultiple(int24 newTick, int24 tickSpacing) internal pure returns (int24) {
+     return (newTick / tickSpacing + (newTick % tickSpacing >= tickSpacing/2 ? int24(1) : int24(0)) ) * tickSpacing;
+    }
+
+    /// @notice Set new optimimal tick range based on current tick
+    /// @param currentFixedRateWad currentFixedRate which is passed in from a 7-day rolling avg. historical fixed rate.
+    function rebalance (int256 currentFixedRateWad) public returns (int256 newTickLower, int256 newTickUpper) {
```
Comment (0xZenus) [inline contracts/strategies/LPOptimiserStrategy.sol]: the current fixed rate cannot be negative. Hence, its type to uint256. Moreover, the ticks need to be returned as int24.

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 10 above)
Comment (0xZenus) [reply contracts/strategies/LPOptimiserStrategy.sol]: It'd be also good to UTs where `current fixed rate -  sigma` is below 0.001 (the lowest fixed rate supported by our vamm) and also where `current fixed rate +  sigma` is above 1000. 

The more cases for UTs, the better in these scenarios.

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 11 above)
Comment (0xZenus) [reply contracts/strategies/LPOptimiserStrategy.sol]: Agree. The fixed rates can't be negative.

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 12 above)
Comment (0xZenus) [reply contracts/strategies/LPOptimiserStrategy.sol]: Agree. I think we should control this via bounds.

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 13 above)
Comment (0xZenus) [reply contracts/strategies/LPOptimiserStrategy.sol]: In that code snippet from voltz-core, we convert sqrtPrice to fixed rate. Here we need to convert tick to fixed rate.

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 14 above)
Comment (0xZenus) [reply contracts/strategies/LPOptimiserStrategy.sol]: Agree, and also the local variables shouldn't have _ in front. Only smart contract variables should have that.

File: contracts/strategies/LPStrategyIndex.ts (hunk 15 above)
Comment (Sol3iman) [reply contracts/strategies/LPStrategyIndex.ts:1]: Yes, this will be the script in ts. Will need to convert to js before placing it into defender.

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 17]
```diff
@@ -43,11 +57,133 @@ contract LPOptimiserStrategy is DefaultAccessControl, ILpCallback {
...
+            return (_newTickLowerMul, _newTickUpperMul);
+        } else {
+            revert(ExceptionsLibrary.REBALANCE_NOT_NEEDED);
+          }
+        }
+
+
+//--------------------// Don't need the below functions // ---------------------//
+
     /// @notice Get new signal and act according to it
```
Comment (0xZenus) [inline contracts/strategies/LPOptimiserStrategy.sol]: need to remove all functions downwards. they are trading vault specific

Comment (0xZenus) [reply contracts/strategies/LPOptimiserStrategy.sol]: also remove ILPCallback inheritance after you do this.

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 18]
```diff
@@ -43,11 +57,133 @@ contract LPOptimiserStrategy is DefaultAccessControl, ILpCallback {
         _erc20Vault = erc20vault_;
         _vault = vault_;
         _marginEngine = IMarginEngine(vault_.marginEngine());
+        _periphery = IPeriphery(vault_.periphery());
+        _vamm = IVAMM(vault_.vamm());
         _currentPosition = vault_.currentPosition();
         _tokens = vault_.vaultTokens();
         _pullExistentials = vault_.pullExistentials();
```
Comment (0xZenus) [inline contracts/strategies/LPOptimiserStrategy.sol:97]: We'll need an event here that captures strategy inception.

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 19]
```diff
@@ -43,11 +57,133 @@ contract LPOptimiserStrategy is DefaultAccessControl, ILpCallback {
...
+            console.logInt(_newTickUpper);
+
+            int24 _newTickLowerMul = nearestTickMultiple(int24(_newTickLower), _tickSpacing);
+            int24 _newTickUpperMul = nearestTickMultiple(int24(_newTickUpper), _tickSpacing);
+
+            console.logInt(_newTickLowerMul);
+            console.logInt(_newTickUpperMul);
+
+            emit Rebalanced(_newTickLowerMul, _newTickUpperMul);
+            return (_newTickLowerMul, _newTickUpperMul);
```
Comment (0xZenus) [inline contracts/strategies/LPOptimiserStrategy.sol]: I see that these are calculated but the global lower and upper ticks of the position are not refreshed. Plus, there should also be a call to vault rebalance.

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 20]
```diff
@@ -43,11 +57,133 @@ contract LPOptimiserStrategy is DefaultAccessControl, ILpCallback {
...
+
+    function setConstants(int24 logProx, uint256 sigmaWad, uint256 max_possible_lower_bound, int24 tickSpacing) public {
+        _requireAtLeastOperator();
+        _logProximity = logProx;
+        _sigmaWad = sigmaWad;
+        _max_possible_lower_bound = max_possible_lower_bound;
+        _tickSpacing = tickSpacing;
+    }
+
+    function setTickValues(int24 currentTick, int24 tickLower, int24 tickUpper) public {
```
Comment (0xZenus) [inline contracts/strategies/LPOptimiserStrategy.sol]: I don't see where this function is going to be used.

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 21]
```diff
@@ -43,11 +57,133 @@ contract LPOptimiserStrategy is DefaultAccessControl, ILpCallback {
...
+        // int24 _currentTick = _periphery.getCurrentTick(_periphery.address());
+
+        // 3. Compare current fixed rate to lower and upper bounds
+        if (
+            _tickLower - _logProximity <= _currentTick &&
+            _currentTick <= _tickUpper + _logProximity
+            
+            ) {
+            // 4.1. If current fixed rate is within bounds, return false (don't rebalance)
+            return (false);
```
Comment (0xZenus) [inline contracts/strategies/LPOptimiserStrategy.sol]: you can just return `false`. No need to use brackets.

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 22]
```diff
@@ -43,11 +57,133 @@ contract LPOptimiserStrategy is DefaultAccessControl, ILpCallback {
...
+        } else {
+            // 4.2. If current fixed rate is outside bounds, return true (do rebalance)
+            return (true);
```
Comment (0xZenus) [inline contracts/strategies/LPOptimiserStrategy.sol]: same.

**Review by 0xZenus — CHANGES_REQUESTED** [review]

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 6 above)
Comment (Sol3iman) [reply contracts/strategies/LPOptimiserStrategy.sol]: That's right. I corrected this.

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 1 above)
Comment (Sol3iman) [reply contracts/strategies/LPOptimiserStrategy.sol:16]: This is needed as prb math is used in the contract

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 3 above)
Comment (Sol3iman) [reply contracts/strategies/LPOptimiserStrategy.sol]: Agreed. I have added wad to the max_possible_lower_bound variable. Not sure whether k_unwind is going to be supplied in wad or not?

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 4 above)
Comment (Sol3iman) [reply contracts/strategies/LPOptimiserStrategy.sol]: That's right.

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 14 above)
Comment (Sol3iman) [reply contracts/strategies/LPOptimiserStrategy.sol]: I think I have addressed this issue. Let's review and check which variables I missed.

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 17 above)
Comment (Sol3iman) [reply contracts/strategies/LPOptimiserStrategy.sol]: I have removed these

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 16 above)
Comment (Sol3iman) [reply contracts/strategies/LPOptimiserStrategy.sol]: I have addressed this

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 20 above)
Comment (Sol3iman) [reply contracts/strategies/LPOptimiserStrategy.sol]: I wrote this function so I could set the required tick values in the tests. Effectively, once everything is wired up to the vault when merged this function would become obsolete.

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 18 above)
Comment (Sol3iman) [reply contracts/strategies/LPOptimiserStrategy.sol:97]: Now there is an event called StrategyDeployment

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 19 above)
Comment (Sol3iman) [reply contracts/strategies/LPOptimiserStrategy.sol]: The global variables are now updated

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 21 above)
Comment (Sol3iman) [reply contracts/strategies/LPOptimiserStrategy.sol]: Fixed

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 22 above)
Comment (Sol3iman) [reply contracts/strategies/LPOptimiserStrategy.sol]: Fixed

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 23]
```diff
@@ -25,12 +27,30 @@ contract LPOptimiserStrategy is DefaultAccessControl, ILpCallback {
...
     IMarginEngine internal _marginEngine;
+    IPeriphery internal _periphery;
+    IVAMM internal _vamm;
     IVoltzVault.TickRange internal _currentPosition;
 
     // MUTABLE PARAMS
 
     uint256 _lastSignal;
     uint256 _lastLeverage;
+    uint256 _k_unwind_parameter; // parameter for k*leverage (for unwinding so this needs to be sent to the contract vault but not used in the strategy vault)
```
Comment (0xZenus) [inline contracts/strategies/LPOptimiserStrategy.sol]: _lastSignal, _lastLeverage -- not used at all. These were used for trading strategy.

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 24]
```diff
@@ -25,12 +27,30 @@ contract LPOptimiserStrategy is DefaultAccessControl, ILpCallback {
...
+    uint256 _sigmaWad; // y (standard deviation parameter in wad 10^18)
+    int256 _max_possible_lower_bound_wad; // should be in fixed rate
+
+    int24 _logProximity; // x (closeness parameter in wad 10^18) in log base 1.0001
+    int24 _currentTick;
+    int24 _tickLower;
+    int24 _tickUpper;
+    int24 _tickSpacing;
```
Comment (0xZenus) [inline contracts/strategies/LPOptimiserStrategy.sol]: It's good to explicitly mention the modifier for each variable. For example, these should be internal. Moreover, variables as sigma wad, max bound, log proximity should have independent setters and getters.

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 25]
```diff
@@ -25,12 +27,30 @@ contract LPOptimiserStrategy is DefaultAccessControl, ILpCallback {
...
     uint256 _lastSignal;
     uint256 _lastLeverage;
+    uint256 _k_unwind_parameter; // parameter for k*leverage (for unwinding so this needs to be sent to the contract vault but not used in the strategy vault)
+    uint256 _sigmaWad; // y (standard deviation parameter in wad 10^18)
+    int256 _max_possible_lower_bound_wad; // should be in fixed rate
+
+    int24 _logProximity; // x (closeness parameter in wad 10^18) in log base 1.0001
+    int24 _currentTick;
+    int24 _tickLower;
+    int24 _tickUpper;
```
Comment (0xZenus) [inline contracts/strategies/LPOptimiserStrategy.sol]: I don't see why we store current tick, lower tick or upper tick here. We should retrieve them directly from the vault.

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 26]
```diff
@@ -25,12 +27,30 @@ contract LPOptimiserStrategy is DefaultAccessControl, ILpCallback {
     IVoltzVault internal _vault;
     uint256[] internal _pullExistentials;
     IMarginEngine internal _marginEngine;
+    IPeriphery internal _periphery;
+    IVAMM internal _vamm;
     IVoltzVault.TickRange internal _currentPosition;
```
Comment (0xZenus) [inline contracts/strategies/LPOptimiserStrategy.sol:32]: I don't see why we need to keep track of _currentPosition here. We should retrieve it directly from the vault.

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 27]
```diff
@@ -43,135 +63,109 @@ contract LPOptimiserStrategy is DefaultAccessControl, ILpCallback {
...
-    /// @param marginUpdate This flags that this update is triggered by deposit/withdrawal
-    function update(
-        uint256 signal, 
-        uint256 leverage,
-        bool marginUpdate
-    ) public {
+        emit StrategyDeployment(erc20vault_, vault_, admin_);
+    }
+    
+    function setConstants(int24 logProx, uint256 sigmaWad, int256 max_possible_lower_bound_wad, int24 tickSpacing) public {
```
Comment (0xZenus) [inline contracts/strategies/LPOptimiserStrategy.sol]: I understand if you use this function for tests. However, variables should have their independent setter/getter.

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 28]
```diff
@@ -43,135 +63,109 @@ contract LPOptimiserStrategy is DefaultAccessControl, ILpCallback {
...
+    /// @notice Set new optimimal tick range based on current tick
+    /// @param currentFixedRateWad currentFixedRate which is passed in from a 7-day rolling avg. historical fixed rate.
+    function rebalance (uint256 currentFixedRateWad) public returns (int24 newTickLowerMul, int24 newTickUpperMul) {
+        _requireAtLeastOperator();
+
+            // 0. Get tickspacing from vamm
+            // _tickSpacing = _vamm.tickSpacing(_vamm);
+
+            // 1. Get the new tick lower
+            int256 deltaWad = int256(currentFixedRateWad - _sigmaWad);
```
Comment (0xZenus) [inline contracts/strategies/LPOptimiserStrategy.sol]: Edge case here: What if _sigmaWad > currentFixedRateWad? It underflows in this situation. You should consider int256(currentFixedRateWad) - int256(_sigmaWad).

You should write a test that captures this scenario.

**Review by 0xZenus — CHANGES_REQUESTED** [review]

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 28 above)
Comment (Sol3iman) [reply contracts/strategies/LPOptimiserStrategy.sol]: Modified this and added tests

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 27 above)
Comment (Sol3iman) [reply contracts/strategies/LPOptimiserStrategy.sol]: Added setters and getters

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 26 above)
Comment (Sol3iman) [reply contracts/strategies/LPOptimiserStrategy.sol:32]: Removed

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 25 above)
Comment (Sol3iman) [reply contracts/strategies/LPOptimiserStrategy.sol]: Ticks are being retrieved from vault and these were removed

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 24 above)
Comment (Sol3iman) [reply contracts/strategies/LPOptimiserStrategy.sol]: These are all declared as internal now

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 23 above)
Comment (Sol3iman) [reply contracts/strategies/LPOptimiserStrategy.sol]: These are removed now
